server/src/main/java/org/apache/druid/metadata/SQLMetadataConnector.java

```diff
@@ -350,6 +381,87 @@ tableName, getPayloadType(), getCollation()
     );
   }
 
+  public void alterEntryTable(final String tableName)
+  {
+    try {
+      retryWithHandle(
+          new HandleCallback<Void>()
+          {
+            @Override
+            public Void withHandle(Handle handle)
+            {
+              if (!tableContainsColumn(handle, tableName, "type")) {
+                log.info("Adding column: type to table[%s]", tableName);
+                handle.execute(StringUtils.format("ALTER TABLE %1$s ADD COLUMN type VARCHAR(255)", tableName));
+              }
+              if (!tableContainsColumn(handle, tableName, "group_id")) {
+                log.info("Adding column: group_id to table[%s]", tableName);
+                handle.execute(StringUtils.format("ALTER TABLE %1$s ADD COLUMN group_id VARCHAR(255)", tableName));
+              }
+              return null;
+            }
+          }
+      );
+    }
+    catch (Exception e) {
+      log.warn(e, "Exception altering table");
+    }
+  }
+
+  @Override
+  public boolean migrateTaskTable()
+  {
+    final MetadataStorageTablesConfig tablesConfig = tablesConfigSupplier.get();
+    final String entryType = tablesConfig.getTaskEntryType();
+    final String tableName = tablesConfig.getEntryTable(entryType);
+    return migrateTaskTable(tableName);
+  }
+
+  public boolean migrateTaskTable(String tableName)
+  {
+    log.info("Populate fields task and group_id of task entry table [%s] from payload", tableName);
+    try {
+      retryWithHandle(
+          new HandleCallback<Void>()
+          {
+            @Override
+            public Void withHandle(Handle handle) throws SQLException, IOException
+            {
+              ObjectMapper objectMapper = new ObjectMapper();
+              Connection connection = handle.getConnection();
+              Statement statement = connection.createStatement(ResultSet.TYPE_SCROLL_INSENSITIVE, ResultSet.CONCUR_UPDATABLE);
+              boolean flag = true;
+              while (flag) {
+                // Should ideally use a cursor and sort by id for efficiency, but updates with ordering aren't allowed
+                String sql = StringUtils.format(
+                    "SELECT * FROM %1$s WHERE active = false AND type IS null %2$s",
+                    tableName,
+                    limitClause(100)
+                );
+                ResultSet resultSet = statement.executeQuery(sql);
+                flag = false;
+                while (resultSet.next()) {
+                  ObjectNode payload = objectMapper.readValue(resultSet.getBytes("payload"), ObjectNode.class);
+                  resultSet.updateString("type", payload.get("type").asText());
+                  resultSet.updateString("group_id", payload.get("groupId").asText());
+                  resultSet.updateRow();
```

Comment:
Doing updates like this in the same transaction is overkill and might get into weird locking behaviors on the table.  It's best not to do it.  Each update should be effectively 2 queries: 1 to get the task payloads, that would exit and return all DB resources and then a second one to actually issue an UPDATE command to set the extra fields based on the taskId.

When the first query stops returning results, that's when you know the migration is complete.

Reply:
Fwiw, you probably don't even need to set a "limit" clause on the query anymore if you follow what I suggested above.  Basically, if you just tell JDBI to only give you the first 100 things from the ResultSet, it should be able to just return them and then when you close the cursor, the DB will close things and be happy-ish.

Reply:
Separated read and update into separate transactions during migration and resources are released after each iteration

Reply:
Added logs and a sleep after each iteration